arch/xtensa/core/mpu.c

```diff
@@ -701,3 +707,419 @@ void xtensa_mpu_init(void)
 	/* Write the map into hardware. There is no turning back now. */
 	xtensa_mpu_map_write(&xtensa_mpu_map_fg_kernel);
 }
+
+#ifdef CONFIG_USERSPACE
+
+/**
+ * @brief Decode kernel_permission for R/W/X.
+ *
+ * This decodes the kernel part of the permission to extract whether
+ * it indicates readable, writable and executable.
+ *
+ * @param[in]  perm Permission to be decoded.
+ * @param[out] readable True if permission is kernel readable.
+ * @param[out] writable True if permission is kernel writable.
+ * @param[out] executable True if permission is kernel executable.
+ */
+static void decode_kernel_permission(uint32_t perm, bool *readable,
+				     bool *writable, bool *executable)
```

Comment:
So, this is a lot of code.  And I guess I don't understand why it's needed?  We shouldn't ever need to decode a MPU entry, those bits only ever get inspected by the hardware.  The kernel's job is unidirectional: we need to translate "Zephyr Permissions" to the MPU bits only.  If code in Zephyr needs to know what the permissions are for a given memory region, it can just ask the memory_domain, which is what it's for.

Reply:
This is mainly used to figure out what kernel permissions are so we can remove the user part of permissions when a partition is removed from the memory domain. Some other arch defaults back to kernel R/W. I just want to go further to make sure no additional permissions are granted.

Anyway, I have converted it into switch/case.